### ISSUE TITLE
VMap IllegalStateException: Failed to find a hash

### ISSUE THREAD

**Issue by lpetrovic05** [hashgraph/hedera-services#8236]
Test data: http://35.247.76.217:8095/swirlds-automation/develop/6N/Migration/20230828-091035-GCP-Daily-Migration-6N/FCM-Migration-NewNodesFromSavedState-2.5k-10m/

```
2023-08-28 09:25:42.999 168      INFO  STATE_TO_DISK    <<platform-core: signed-state-file-manager 5>> SignedStateFileWriter: Finished writing state for round 17296 to disk. Reason: PERIODIC_SNAPSHOT, directory: /opt/hgcapp/services-hedera/HapiApp2.0/data/saved/com.swirlds.demo.migration.MigrationTestingToolMain/5/123/17296 {"round":17296,"freezeState":false,"reason":"PERIODIC_SNAPSHOT","directory":"file:///opt/hgcapp/services-hedera/HapiApp2.0/data/saved/com.swirlds.demo.migration.MigrationTestingToolMain/5/123/17296/"} [com.swirlds.logging.payloads.StateSavedToDiskPayload]
2023-08-28 09:25:57.781 316      FATAL EXCEPTION        <<platform-core: state-hash-sign 5>> ManualWiring: Fatal error, node will shut down. Reason: Exception occurred during SignedState hashing {} [com.swirlds.logging.payloads.FatalErrorPayload]
   java.base/java.lang.Thread.getStackTrace(Thread.java:1610)
   com.swirlds.platform.components.wiring.ManualWiring.logFatalError(ManualWiring.java:235)
   com.swirlds.platform.components.wiring.ManualWiring.handleFatalError(ManualWiring.java:213)
   com.swirlds.platform.state.signed.SignedStateHasher.hashState(SignedStateHasher.java:95)
   com.swirlds.platform.components.state.DefaultStateManagementComponent.newSignedStateFromTransactions(DefaultStateManagementComponent.java:381)
   com.swirlds.common.threading.framework.internal.QueueThreadImpl.doWork(QueueThreadImpl.java:253)
   com.swirlds.common.threading.framework.internal.StoppableThreadImpl.doWork(StoppableThreadImpl.java:614)
   com.swirlds.common.threading.framework.internal.StoppableThreadImpl.run(StoppableThreadImpl.java:215)
   java.base/java.lang.Thread.run(Thread.java:833)

java.util.concurrent.ExecutionException: java.lang.RuntimeException: exception encountered while hashing virtual tree
	at com.swirlds.common.merkle.hash.FutureMerkleHash.rethrowException(FutureMerkleHash.java:83) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.merkle.hash.FutureMerkleHash.get(FutureMerkleHash.java:93) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.merkle.hash.FutureMerkleHash.get(FutureMerkleHash.java:30) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.platform.state.signed.SignedStateHasher.hashState(SignedStateHasher.java:84) ~[swirlds-platform-core-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.platform.components.state.DefaultStateManagementComponent.newSignedStateFromTransactions(DefaultStateManagementComponent.java:381) ~[swirlds-platform-core-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.threading.framework.internal.QueueThreadImpl.doWork(QueueThreadImpl.java:253) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.threading.framework.internal.StoppableThreadImpl.doWork(StoppableThreadImpl.java:614) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.threading.framework.internal.StoppableThreadImpl.run(StoppableThreadImpl.java:215) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.RuntimeException: exception encountered while hashing virtual tree
	at com.swirlds.virtualmap.internal.hash.VirtualHasher.hashSubTree(VirtualHasher.java:708) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.virtualmap.internal.hash.VirtualHasher.hash(VirtualHasher.java:424) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.virtualmap.internal.merkle.VirtualRootNode.computeHash(VirtualRootNode.java:1264) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.virtualmap.internal.pipeline.VirtualPipeline.hashCopy(VirtualPipeline.java:388) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.virtualmap.internal.merkle.VirtualRootNode.getHash(VirtualRootNode.java:1204) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.merkle.hash.MerkleHashBuilder.hashSubtree(MerkleHashBuilder.java:206) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at com.swirlds.common.merkle.hash.MerkleHashBuilder.lambda$createHashingRunnable$1(MerkleHashBuilder.java:173) ~[swirlds-common-0.42.0-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	... 1 more
Caused by: java.lang.IllegalStateException: Failed to find a hash for 66548, which is a sibling of 66547
	at com.swirlds.virtualmap.internal.hash.VirtualHasher.lambda$hashSubTree$2(VirtualHasher.java:667) ~[swirlds-virtualmap-0.42.0-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
```

**Comment by imalygin** [comment]
What we know about this bug so far:
1. It reproduces regularly
2. It's non-deterministic. It happens on creating snapshot. However, it may create a couple of snapshots successfully and then fail. 
3. It's highly unlikely that it's caused by the persisted state the node starts with. 
4. It happens because `VitualNodeHasher` is not able to get a hash from `VirtualNodeCache` - method `VirtualNodeCache#lookupHashByPath` returns null because `pathToDirtyHashIndex` doesn't contain a hash for this path, even though it was expected to be there
5. This happened because previously `VirtualNodeCache#dirtyHashesForFlush` did not return an instance of `VirtualHashRecord` for this path, even though it had to because `pathToDirtyLeafIndex` actually contains a record for this path. 
6. This happened because a mutation for the investigated path is absent.  At this point it is not clear why it is the case. 
7. To make sure that this issue was introduced in 0.42 I tried to run the same test against the same state - unfortunately, it didn't work out - the state is not backward compatible.

**Comment by imalygin** [comment]
We had a long debug session (@OlegMazurov, @artemananiev and myself) and discovered that the virtual cache behavior was really weird - `VirtualNodeCache#lookupLeafByPath` method returned an instance of `VirtualLeafRecord` that had a that is not equal to that path passed as a parameter. 

Here is another bit of a strange behavior - we could find a virtual record by a path from a valid range `[firstLeafPath, lastLeafPath]`, however when we used a key from the record to look up this record. the datasource returned `null`. 

At this point we got convinced that the persisted state is to blame. However, it was not clear how exactly it got corrupted in a way that it works only partially, but it entirely broken. The state files looked correct, there was nothing obvious like a file of a zero size when it's not supposed to be empty. 

We validated `equals` and `hashCode` of `com.swirlds.demo.migration.virtual.AccountVirtualMapKey` that is used as a a key in this virtual record - it is implemented correctly. However, we didn't anticipate the fact that the `hashCode` implementation has changed recently. Even though both implementations are correct, they are _different_ ([this line specifically](https://github.com/hashgraph/hedera-services/pull/8015/files#diff-a389c758e48144632359c50e9a36849dc8e06346fcaf00a1fd73726be5703d86R150)( which ultimately caused the incompatibility.
